animdustry: init at 1.2-unstable-2025-09-18

Ezjfc opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [X] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [X] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ezjfc committed:
animdustry: init at 1.2-unstable-2025-09-18

RossSmyth reviewed (changes requested):
Thanks for the contribution!

Ezjfc commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,122 @@
+      created as a mindustry april 1st event";
+    '';
+    sourceProvenance = with sourceTypes; [ fromSource ];
+    license = licenses.gpl3;
```
Hi, the license file was listed in animdustry.nimble
https://github.com/Anuken/animdustry/blob/f408e632872929964a9b3f8888f1c7a18e6c1ead/animdustry.nimble#L4

```suggestion
    license = licenses.gpl3; # Listed in animdustry.nimble
```

RossSmyth replied:
I see, alright.

Ezjfc commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,122 @@
+      categories = [ "Game" ];
+    };
+
+  meta = with lib; {
```
Seems like it is still needed by `    maintainers = with maintainers; [` at line 118.

Sigmanificient replied:
You want to use `maintainers = with lib.maintainers; [`

Ezjfc commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,122 @@
+    runHook preInstall
+
+    mv $out/bin/main $out/bin/animdustry
+    wrapProgram $out/bin/animdustry \
+      --suffix LD_LIBRARY_PATH : ${
+        lib.makeLibraryPath [
+          libpulseaudio
+        ]
+      }
```
Thank you for the suggestion. I will take a look.

Ezjfc replied:
Sorry, patchelf did not work well for it. I am getting errors like this:
```
nix-configurations/pkgs on  wip [✘»!+?] took 59s 
❯ ./result/bin/animdustry
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: no version information available (required by /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry)
/home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: symbol lookup error: /home/rickastley/Documents/nix/nix-configurations/pkgs/./result/bin/animdustry: undefined symbol: , version 
```

when I build it with
```bash
    patchelf --set-interpreter ${stdenv.cc.bintools.dynamicLinker} $BIN
    patchelf --set-rpath ${
      lib.makeLibraryPath (
        buildInputs ++ [ libpulseaudio ]
      )
    }:$(patchelf --print-rpath $BIN) $BIN
```

Using `autoPatchelfHook` also did not work as `libpulseaudio` required by `animdustry` could not be detected by `ldd`.

Ezjfc replied:
So I would like to keep the current `wrapProgram` way.

RossSmyth replied:
This works:
```patch
diff --git a/pkgs/by-name/an/animdustry/package.nix b/pkgs/by-name/an/animdustry/package.nix
index d29f99626e..d29261524e 100644
--- a/pkgs/by-name/an/animdustry/package.nix
+++ b/pkgs/by-name/an/animdustry/package.nix
@@ -2,13 +2,11 @@
   lib,
   copyDesktopItems,
   makeDesktopItem,
-  makeBinaryWrapper,
+  autoPatchelfHook,
   writeText,
-
   fetchFromGitHub,
   buildNimPackage,
   nim-1_0,
-
   xorg,
   libGL,
   libGLU,
@@ -64,9 +62,8 @@
   ];
   nativeBuildInputs = [
     faupack
-
-    makeWrapper
     copyDesktopItems
+    autoPatchelfHook
   ];
 
   nimFlags = [
@@ -77,16 +74,14 @@
     "-d:NimblePkgVersion=${finalAttrs.version}"
   ];
 
+  runtimeDependencies = [
+    libpulseaudio
+  ];
+
   installPhase = ''
     runHook preInstall
 
     mv $out/bin/main $out/bin/animdustry
-    wrapProgram $out/bin/animdustry \
-      --suffix LD_LIBRARY_PATH : ${
-        lib.makeLibraryPath [
-          libpulseaudio
-        ]
-      }
     install -Dm644 ./assets/icon.png $out/share/icons/hicolor/64x64/apps/animdustry.png
 
     runHook postInstall
```

Ezjfc replied:
I also added `stdenv.cc.cc.lib` for:
```
       >     libstdc++.so.6 -> not found!
       >     libgcc_s.so.1 -> not found!
```

Ezjfc commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,122 @@
+  libGLU,
+  libpulseaudio,
+}:
+let
+  buildNimPackage' = buildNimPackage.override {
+    nim2 = nim-1_0; # Intended for Nim 1, but resolves to Nim 2 because of "nim >= 1.6.2" in animdustry.nimble
+  };
+
+  fau = fetchFromGitHub {
+    owner = "Anuken";
+    repo = "fau";
+    rev = "73df4a699873d0f82fd612a2a2ac63c21d3f2233";
+    hash = "sha256-9zwmFinDJV4+R/aiVVOQ/Bv30jX7NHJyufzMNWHGA+k=";
+    fetchSubmodules = true;
+  };
+  faupack = buildNimPackage' {
+    pname = "fau";
+    version = "0-unstable-2022-05-14";
+    src = fau;
+    lockFile = ./fau-lock.json;
+  };
+
+  fauLock = builtins.fromJSON (builtins.readFile ./fau-lock.json);
+  partialLock = builtins.fromJSON (builtins.readFile ./package-lock-partial.json);
+  lockFile = writeText "package-lock.json" (builtins.toJSON {
+    depends = fauLock.depends ++ partialLock.depends;
+  });
+in
```
The `fau` source code and its dependencies are required when building `animdustry`:
https://github.com/NixOS/nixpkgs/blob/dde1ed711b06e7290dcb6ac721ea43f4f99a61e5/pkgs/by-name/an/animdustry/package.nix#L75-L76

If this suggestion is to make the code cleaner, it might not be worth it.

Ezjfc replied:
Please consider this change, and if you are still in favour of this change, I will do it.

RossSmyth replied:
Yeah, you can just do `fau.src`

Sigmanificient commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,121 @@
+  };
+
+  buildInputs = [
+    xorg.libX11
+    xorg.libXcursor
+    xorg.libXrandr
+    xorg.libXinerama
+    xorg.libXi
```
Theses can be imported without the `xorg` namespace

Sigmanificient commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,121 @@
+  src = fetchFromGitHub {
+    owner = "Anuken";
+    repo = "animdustry";
+    rev = "v${finalAttrs.version}";
```
```suggestion
    tag = "v${finalAttrs.version}";
```

Ezjfc commented:
Sorry, this is my first pkg submission and I have this question in mind:

RossSmyth commented:
This point is providing the libraries that can be loaded.

RossSmyth commented:
<img width="319" height="116" alt="image" src="https://github.com/user-attachments/assets/f86aecb0-1142-430a-b9b8-db38594ac347" />
<img width="515" height="48" alt="image" src="https://github.com/user-attachments/assets/238a2dd0-dae7-4856-97f1-061b2f9a242f" />

Sigmanificient commented:
Your derivation is doing a `IFD` (Import from derivation) which need to be solved in order to pass the CI

RossSmyth commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,115 @@
+  libpulseaudio,
+}:
+let
+  buildNimPackage' = buildNimPackage.override {
+    nim2 = nim-1_0; # Intended for Nim 1, but resolves to Nim 2 because of "nim >= 1.6.2" in animdustry.nimble
+  };
```
Looking closer, it is probably a good idea to use the latest commits that support Nim 2 and not the latest release.
```suggestion
```

Ezjfc replied:
Please see: https://github.com/NixOS/nixpkgs/pull/436806#issuecomment-3235014086

RossSmyth commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,115 @@
+      created as a mindustry april 1st event
+    '';
+    license = lib.licenses.gpl3; # Listed in animdustry.nimble
+    platforms = lib.platforms.all;
```
Seems to only support windows, linux, mac, and android

RossSmyth reviewed (changes requested):
Looking closer at upstream, I think it is worth using later commits rather than just the latest release. 

I also would recommend getting CI green.

normalcea commented:
@RossSmyth 

> Upstream declaring it "done" doesn't really bother me, there are many packages like that. It's not really unmaintained as the author has pushed several commits since the "final" release. If I were packaging this I would use the latest commits and not the latest release.

Using the latest revision would be better then. I'd still consider opening a PR to upstream a desktop file since creating our own is not ideal. 

In any case, the licensing information for the game's assets should be clarified since the GPL licensing is meant for the game's code. If we can't determine a license then `unfree` has to be set.

Ezjfc commented:
@normalcea Hi, although it was stated in the release that the game received its final version, there are still several fixes and improvements, like how @RossSmyth said, such as ["Fau update (major performance fix)"(https://github.com/Anuken/animdustry/commit/6c65e53edf08267865e7d4afced9cfd43dc8dce3)] and [support for Nim 2](https://github.com/Anuken/animdustry/commit/46802f3f4db126c69b582a88f962f68b36a73902). Moreover, the game framework `fau` is still being [updated constantly](https://github.com/Anuken/fau/commits/master/).

However, as those commits were not tagged, instead of selecting the latest commit as @RossSmyth's idea. I would rather register an init stable version (`1.2`) into Nixpkgs before other unstable ones (for example, `1.2-unstable-30-07-2024`). Especially when the author did not publish a precompiled Linux binary for that release:
> no linux build because you can probably build it yourself and I don't want to set up a container right now

Ezjfc commented on pkgs/by-name/an/animdustry/fau.nix:
This file was meant to serve as a helper that provides both the `fau` source and derivation as inputs for Animdustry. However, I couldn't find a way to extract the `src` attribute directly from the derivation. If you have any insights on this, I would appreciate it.

Ezjfc replied:
Hi, I have addressed this problem.

otherthanemry commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,115 @@
+}:
+let
+  buildNimPackage' = buildNimPackage.override {
+    nim2 = nim-1_0; # Intended for Nim 1, but resolves to Nim 2 because of "nim >= 1.6.2" in animdustry.nimble
```
Sorry that `buildNimPackage` isn't better documented, use this instead:
```
buildNimPackage {
  requiredNimVersion = 1;
}
```

otherthanemry commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,115 @@
+  );
+in
+buildNimPackage' (finalAttrs: {
+  inherit lockFile;
```
Probably better build fau and animdustry with the same merged lockfile. Extra dependencies shouldn't increase the runtime closure.
Put all the Nix code into one `package.nix` file to reduce the attack surface for reviewers.

Ezjfc commented:
Sorry for the two unverified commits, I forgot to sign them when cherry-picking.

Ezjfc commented:
@RossSmyth Hi, I kindly request permission to force-push onto this pull request in order to fix my GPG signature.

normalcea commented:
**review** (blocking): Move fau subpackage into `package.nix`

Since fau.nix is uncomplicated, we can include it as a subpackage within animdustry's package.nix like so:

``` diff
diff --git a/pkgs/by-name/an/animdustry/fau.nix b/pkgs/by-name/an/animdustry/fau.nix
deleted file mode 100644
index 8066f261f399..000000000000
--- a/pkgs/by-name/an/animdustry/fau.nix
+++ /dev/null
@@ -1,19 +0,0 @@
-{
-  fetchFromGitHub,
-  buildNimPackage,
-}:
-buildNimPackage {
-  pname = "fau";
-  version = "0-unstable-2022-05-14";
-
-  src = fetchFromGitHub {
-    owner = "Anuken";
-    repo = "fau";
-    rev = "73df4a699873d0f82fd612a2a2ac63c21d3f2233";
-    hash = "sha256-9zwmFinDJV4+R/aiVVOQ/Bv30jX7NHJyufzMNWHGA+k=";
-    fetchSubmodules = true;
-  };
-
-  requiredNimVersion = 1;
-  lockFile = ./lock.json;
-}
diff --git a/pkgs/by-name/an/animdustry/package.nix b/pkgs/by-name/an/animdustry/package.nix
index 0dd70cf552c3..e7371fbdbe17 100644
--- a/pkgs/by-name/an/animdustry/package.nix
+++ b/pkgs/by-name/an/animdustry/package.nix
@@ -32,6 +32,22 @@ buildNimPackage (finalAttrs: {
     hash = "sha256-RND5AhlD9uTMg9Koz5fq7WMgJt+Ajiyi6K60WFbF0bg=";
   };

+  fau = buildNimPackage {
+    pname = "fau";
+    version = "0-unstable-2022-05-14";
+
+    src = fetchFromGitHub {
+      owner = "Anuken";
+      repo = "fau";
+      rev = "73df4a699873d0f82fd612a2a2ac63c21d3f2233";
+      hash = "sha256-9zwmFinDJV4+R/aiVVOQ/Bv30jX7NHJyufzMNWHGA+k=";
+      fetchSubmodules = true;
+    };
+
+    requiredNimVersion = 1;
+    lockFile = ./lock.json;
+  };
+
   buildInputs = [
     stdenv.cc.cc.lib
     libX11
@@ -44,8 +60,7 @@ buildNimPackage (finalAttrs: {
   ];

   nativeBuildInputs = [
-    fau
-
+    finalAttrs.fau
     autoPatchelfHook
     copyDesktopItems
   ];

```

**review** (blocking): remove lock.nix

I don't understand why this file exists, we only need `lock.json`

**review** (blocking): use latest commit

I don't see a strong enough reason to init this package at 1.2 when there's a more up-to-date version from the latest master. Writing `1.2-unstable-XX-YY-ZZZ` is fine and we should be using newer versions of the Nim compiler.

**review** (blocking): meta attrset.

The meta attribute has errors in it:

`longDescription` should is not capitalized nor does it end with a period.

`license` attribute can contain multiple licenses in a list, not just unfree. Also, running the game, the credits lists that the music in the game is licensed under unspecified creative commons licenses, those need to be tracked down as well, it's simply not enough to list everything as unfree.

**nitpick** (style): Remove newline separation and ordering.

There is no hard rule about how derivations should be structured, but using whitespaces/newlines as logical separators isn't needed, derivation builders are so simple anyway that there's no need to structure them. A good rule of thumb of mine is to order arguments/attributes in the order of use.

A good example is this:

``` nix
  desktopItems =
    let
      name = "Animdustry";
      exec = "animdustry";
      icon = exec;
    in
    makeDesktopItem {
      inherit name icon exec;
      desktopName = name;
      categories = [ "Game" ];
    };
```

There's no need to use a `let` binding here since it isn't saving as much time as writing out each entry. Duplicated entries are fine in derivations since the goal should be to read continuously from top to bottom without having to backtrack to understand everything.

Sigmanificient reviewed (commented):
Well, it seems that 5 month ago, animdustry was officially abandonned:
- https://github.com/Anuken/animdustry/commit/55e43ca76690396f80c2e600c47eaf580a47471f

Endermanbugzjfc commented on pkgs/by-name/an/animdustry/package.nix:
```diff
@@ -0,0 +1,109 @@
+      The anime gacha bullet hell rhythm game that was created as a Mindustry (2022) April 1st event.
+    '';
+    license = with lib.licenses; [
+      unfree
```
?????????????????????????????????¿ 🛐🛐

Ezjfc committed:
animdustry: revised meta for package

This commit added the `mainProgram` attribute and changed the `unfree` license to `gpl3Only` after https://github.com/Anuken/animdustry/commit/ce2a5b7b58297b1fe9e7df52d429130e281cfc8d

Co-authored-by: Yohann Boniface <edhyjox@gmail.com>
Co-authored-by: EndermanbugZJFC <175898536+Ezjfc@users.noreply.github.com>
